File: 2_Periodo/AED_II/toAluno/fonte/U4_ordenacao/insercao.c
/**
 * Metodo de ordenacao por insercao
 * @author Max do Val Machado
 * @version 2 01/2015
 */
#include "geracao.h"

// Algoritmo de ordenacao
void insertionSort() {
   for (int i = 1; i < n; i++) {
      int tmp = array[i];
      int j = i - 1;

      while ((j >= 0) && (array[j] > tmp)) {
         array[j + 1] = array[j];
         j--;
      }
      array[j + 1] = tmp;

   }
}

int main() {
   srand(time(NULL));
   decrescente();
   //mostrar();		
   clock_t comeco = clock();
   insertionSort();
   clock_t fim = clock();
   double total = (clock() - comeco) / (double)CLOCKS_PER_SEC / 1000.0;

   mostrar();
   printf("Tempo para ordenar: %f ms (%i).", total, isOrdenado());
}

File: 2_Periodo/AED_II/fonte_aula/Ordenação/SELEÇÃO/Insercao.java
public class Insercao{
	public static void main(String[] args){
		int[] array = new int[4];
		array[0] = 2;
		array[1] = 3;
		array[2] = 2;
		array[3] = 1;
		int n = array.length;
		insercao(array,n);
		for(int i = 0; i < n; i++){
			System.out.println(array[i]);
		}// End for
	}// End main()

	public static void insercao(int[] array,int n){
		for(int i = 0; i < n;i++){
			int tmp = array[i];
			int j = (i-1);
			int value = pesquisabinaria(array,0,j,tmp);
			while(j >= value && array[j] > tmp){
				array[j+1] = array[j];
				j--;
			}// End while
			array[j+1] = tmp;
		}// End for
	}// End insercao()

	public static int pesquisabinaria(int[] array,int i,int j,int x){
		int meio = 0;
		while(i <= j){
			meio = (i+j)/2;
			if(x >= array[meio]){
				i = meio+1;
			}// End if
			else if(x < array[meio]){
				j = meio-1;
			}// End else if
		}// End while
		return meio;
	}// End pesquisabinaria()
}// End class Insercao
